### DISCUSSION TITLE
Device-side grouped GEMM

### DISCUSSION THREAD

**Discussion by osayamenja** [NVIDIA/cutlass#1620] [Q&A] [answered]
Hello!

Is there a way to do a device-side grouped (or batched) gemm? That is, my goal is to perform a set of GEMMs with non uniform sizes, flexible constraint as we can pad zeros, inside a _kernel_. I could serially invoke N `cute::cooperative_gemm` calls, but that may not be as optimal as highlighted in this GTC'24 [talk](https://www.nvidia.com/en-us/on-demand/session/gtc24-s61198/).

Any suggestions for actualizing the grouped gemm, preferably for capability >= 70? Thanks.

**Comment by jackkosaian** [comment]
By "device-side," do you mean that you'd like to perform the grouped-GEMM from within an existing kernel, rather than launching two separate kernels?

**Reply by osayamenja** [reply]
Yes, within a kernel, any way to do this?

**Reply by jackkosaian** [reply]
There is not currently a way to do this.

[cuBLASDx](https://docs.nvidia.com/cuda/cublasdx/) provides some methods for doing this for basic GEMMs. However, I'm not sure that it has been extended to support grouped GEMM.

**Reply by thakkarV** [reply]
Note that CUTLASS collective already support this. There's no such thing as a separate "device side grouped GEMM". That's exactly what the collectives are

**Reply by osayamenja** [reply]
@thakkarV thanks for the suggestion, I had only been looking at `cute` gemm and had no idea about the Collective API in `cutlass`. I'll be doing some reading about that [here](https://github.com/NVIDIA/cutlass/blob/637b15906358191cb4238af419d408a65819d7ec/media/docs/gemm_api_3x.md) :) 

So, achieving this would mean constructing a collective in device code and configuring for grouped like this [example](https://github.com/NVIDIA/cutlass/blob/637b15906358191cb4238af419d408a65819d7ec/examples/57_hopper_grouped_gemm/57_hopper_grouped_gemm.cu), is this similar to what you had in mind?

**Reply by thakkarV** [reply, marked as answer]
yes, that example ultimately uses this collective: https://github.com/NVIDIA/cutlass/blob/main/include/cutlass/gemm/collective/sm90_mma_array_tma_gmma_ss_warpspecialized.hpp

which you can use directly in your code on the device side.

**Reply by osayamenja** [reply]
thanks! @thakkarV @jackkosaian
